libsupport/src/JSONTracer.cpp

```diff
@@ -0,0 +1,385 @@
+}
+
+std::shared_ptr<katana::ProgressSpan>
+katana::JSONTracer::StartSpan(
+    const std::string& span_name, const katana::ProgressContext& child_of) {
+  return JSONSpan::Make(span_name, child_of);
+}
+
+std::string
+katana::JSONTracer::Inject(const katana::ProgressContext& ctx) {
+  return ctx.GetTraceID() + "," + ctx.GetSpanID();
+}
+
+std::unique_ptr<katana::ProgressContext>
+katana::JSONTracer::Extract(const std::string& carrier) {
+  size_t split = carrier.find(",");
+  if (split == std::string::npos) {
+    return nullptr;
+  } else {
+    std::string trace_id = carrier.substr(0, split);
+    std::string span_id = carrier.substr(split + 1);
+    return std::unique_ptr<JSONContext>(new JSONContext(trace_id, span_id));
+  }
+}
+
+std::unique_ptr<katana::ProgressContext>
+katana::JSONContext::Clone() const noexcept {
+  return std::unique_ptr<JSONContext>(new JSONContext(trace_id_, span_id_));
+}
+
+katana::JSONSpan::JSONSpan(
+    const std::string& span_name, std::shared_ptr<katana::ProgressSpan> parent)
+    : ProgressSpan(parent), context_(JSONContext{"", ""}) {
+  auto& tracer = ProgressTracer::GetProgressTracer();
+
+  std::string parent_span_id{"null"};
+  std::string trace_id;
+  std::string host_data;
+  if (parent != nullptr) {
+    auto parent_span = std::static_pointer_cast<JSONSpan>(parent);
+    parent_span_id = parent_span->GetContext().GetSpanID();
+    trace_id = parent_span->GetContext().GetTraceID();
+  } else {
+    trace_id = GenerateID(tracer.GetHostID());
+    host_data = GetHostStatsJSON();
+  }
+  std::string span_id = GenerateID(tracer.GetHostID());
+  context_ = JSONContext(trace_id, span_id);
+
+  std::string message{"start"};
+
+  std::string span_data = GetSpanJSON(span_id, span_name, parent_span_id);
+  std::string log_data = GetLogJSON(message);
+  std::string tag_data;
+
+  std::string output_json =
+      BuildJSON(trace_id, span_data, log_data, tag_data, host_data);
+
+  std::cerr << output_json;
+}
+
+katana::JSONSpan::JSONSpan(
+    const std::string& span_name, const katana::ProgressContext& parent)
+    : ProgressSpan(nullptr), context_(JSONContext{"", ""}) {
+  auto& tracer = ProgressTracer::GetProgressTracer();
+
+  std::string parent_span_id = parent.GetSpanID();
+  std::string trace_id = parent.GetTraceID();
+  std::string span_id = GenerateID(tracer.GetHostID());
+  context_ = JSONContext(trace_id, span_id);
+
+  std::string message{"start"};
+
+  std::string host_data = GetHostStatsJSON();
+  std::string span_data = GetSpanJSON(span_id, span_name, parent_span_id);
+  std::string log_data = GetLogJSON(message);
+  std::string tag_data;
+
+  std::string output_json =
+      BuildJSON(trace_id, span_data, log_data, tag_data, host_data);
+
+  std::cerr << output_json;
+}
+
+std::shared_ptr<katana::ProgressSpan>
+katana::JSONSpan::Make(
+    const std::string& span_name,
+    std::shared_ptr<katana::ProgressSpan> parent) {
+  return std::shared_ptr<JSONSpan>(new JSONSpan(span_name, parent));
```

Comment:
why return a shared_ptr when we allocate a new `JSONSpan` and don't keep a reference ourselves?

Reply:
For consistency since spans can be accessed in various ways

Reply:
If you want to support ease of access, it is better to return the type that says the least about something. unique_ptrs can always be made into shared_ptrs but the converse is not true.

If you want to support a variety of allocation patterns (e.g., object caching), maybe shared_ptr would make sense here.

Do you want to support a variety of allocation patterns or ease of use?

Reply:
I'll try and scope out how it works but it's mostly ease of access:

All spans maintain a shared_ptr reference to their parent so it does not go out of scope before the child
All spans give the user code some form of access to the shared_ptr-either directly or via a ProgressScope
The currently active span allows returning a reference to the object via a shared_ptr and passing the shared_ptr itself as the parent to a newly created child span

Reply:
Also, `std::move(parent)`